Add support for IServiceProviderIsService

davidfowl opened the pull request:
- This optional service lets consumers query to see if a service is resolvable without side effects (not having to explicitly resolve the service).
- Added new spec tests to verify the baseline behavior based on IServiceCollection features.

Fixes #53919

davidfowl committed:
Add support for IServiceProviderIsService
- This optional service lets consumers query to see if a service is resolvable without side effects (not having to explicitly resolve the service).
- Added new spec tests to verify the baseline behavior based on IServiceCollection features.

alistairjevans commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ServiceProviderIsServiceSpecificationTests.cs:
```diff
@@ -0,0 +1,83 @@
+            // Assert
+            Assert.NotNull(serviceProviderIsService);
+            Assert.True(serviceProviderIsService.IsService(typeof(IEnumerable<IFakeService>)));
+            Assert.True(serviceProviderIsService.IsService(typeof(IEnumerable<FakeService>)));
```
Have you considered adding an assert for the open `typeof(IEnumerable<>)` returning false? It should follow the same behaviour as other open generics, not just special-case because it is `IEnumerable`.

davidfowl replied:
Can do

davidfowl committed:
PR feedback, add IEnumerable open generic test

stephentoub commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:
```diff
@@ -441,6 +441,32 @@ public void Add(Type type, ServiceCallSite serviceCallSite)
+
+            if (serviceType.IsConstructedGenericType && serviceType.GetGenericTypeDefinition() is Type genericDefinition)
+            {
+                return genericDefinition == typeof(IEnumerable<>) || _descriptorLookup.ContainsKey(genericDefinition);
```
Can you add a comment as to why IEnumerable is special-cased?

davidfowl replied:
Done

tillig commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/IServiceProviderIsService.cs:
```diff
@@ -0,0 +1,20 @@
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Optional service used to determine if the specified type is resolvable by the <see cref="IServiceProvider"/>.
```
```suggestion
    /// Optional service used to determine if the specified type is available from the <see cref="IServiceProvider"/>.
```

It may be available, but it's not necessarily resolvable.

tillig replied:
(Same change below.)

davidfowl replied:
Done

davidfowl committed:
PR feedback

tillig commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/IServiceProviderIsService.cs:
```diff
@@ -0,0 +1,20 @@
+        /// Determines if the specified service type is available from the <see cref="IServiceProvider"/>.
+        /// </summary>
+        /// <param name="serviceType">An object that specifies the type of service object to test.</param>
+        /// <returns>true if the specified service is a resolvable, false if it is not.</returns>
```
```suggestion
        /// <returns>true if the specified service is a available, false if it is not.</returns>
```

davidfowl committed:
Update src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/IServiceProviderIsService.cs

Co-authored-by: Travis Illig <tillig@paraesthesia.com>

pakrym commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ServiceProviderIsServiceSpecificationTests.cs:
```diff
@@ -0,0 +1,84 @@
+
+namespace Microsoft.Extensions.DependencyInjection.Specification
+{
+    public abstract class ServiceProviderIsServiceSpecificationTests
```
We only have the instrumentation to run spec tests for 3rd party containers enabled for the main test class.

davidfowl replied:
I started that way but moved away from it because it wasn't very clean. I also tried adding a virtual property to enable running these tests but there's no first class xunit way to skip based on that. Do you have a cleaner idea on how to suppress failures?

pakrym replied:
https://github.com/dotnet/runtime/blob/davidfowl/serviceprovider-isservice/src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/StructureMap.cs#L11 ??

davidfowl replied:
I think if we need to do this then the default base class needs to have this feature. This feature is completely optional and if we have it as part of the spec tests then we need to built this skippibility (I made that word up) as part of the default experience.

pakrym replied:
can you use an `if` inside the and a virtual property to skip?
```
public virtual SupportsIsService {get;} = true;


public void IsServiceWorks()
{
if (SupportsIsService)
{
// test code.

}
}
```

davidfowl replied:
This is what I did originally but it felt so bad. You think its better to do it this way?

pakrym replied:
It gives us coverage and 3rd parties a way to discover new features when updating the spec test package.

davidfowl replied:
So is it for us or for 3rd parties? How will they discover it if all of the tests pass and the virtual returns false by default?

pakrym replied:
>So is it for us or for 3rd partie

Both.

>How will they discover it if all of the tests pass and the virtual returns false by default?

We'll turn it on by default.

davidfowl replied:
Done

pakrym commented:
Can you put `Debug.Assert(CanGetService(type))` into `GetService` implementation and run the test suit with that? Make sure we didn't miss anything.

pakrym commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ServiceProviderIsServiceSpecificationTests.cs:
```diff
@@ -0,0 +1,84 @@
+
+namespace Microsoft.Extensions.DependencyInjection.Specification
+{
+    public abstract class ServiceProviderIsServiceSpecificationTests
```
Add `IServiceScopeProvider` and `IServiceProvider` test cases.

davidfowl replied:
> IServiceScopeProvider

IServiceScopeFactory

> IServiceProvider 

Not sure what you mean by this.

pakrym replied:
```
            Assert.True(serviceProviderIsService.IsService(typeof(IServiceScopeProvider)));
            Assert.True(serviceProviderIsService.IsService(typeof(IServiceProvider)));
```

davidfowl replied:
Got it.

pakrym replied:
And `Assert.True(serviceProviderIsService.IsService(typeof(IServiceProviderIsService)));`

davidfowl replied:
Done

davidfowl committed:
Handle built in services as part of IsServce
- Special case built in services as part of the IsService check
- Added a test case

halter73 commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ServiceProviderIsServiceSpecificationTests.cs:
```diff
@@ -0,0 +1,102 @@
+            Assert.NotNull(serviceProviderIsService);
+            Assert.True(serviceProviderIsService.IsService(typeof(IServiceProvider)));
+            Assert.True(serviceProviderIsService.IsService(typeof(IServiceScopeFactory)));
+            Assert.True(serviceProviderIsService.IsService(typeof(IServiceProviderIsService)));
```
😆

halter73 reviewed (approved):
I think I agree with @pakrym that we should put this in DependencyInjectionSpecificationTests with SupportsIsService true by default for greater discoverability when third-party containers update their Specification.Tests dependency.

davidfowl committed:
PR feedback
- Make the tests part of the core DI tests and enable skipping via a property

davidfowl commented:
> I think I agree with @pakrym that we should put this in DependencyInjectionSpecificationTests with SupportsIsService true by default for greater discoverability when third-party containers update their Specification.Tests dependency.

Done!

davidfowl commented:
Errors as unrelated, Merging.
